### ISSUE TITLE
Do mathematics for State Abstraction

### ISSUE THREAD

**Issue by Gauraviitkgp** [Gauraviitkgp/RubikCubeRL#5]

**Comment by Raghwendra-Dey** [comment]
since fixing the reduction matrix for state abstraction didn't seem to work, we could try to make the matrix trainable and check?

**Comment by Gauraviitkgp** [comment]
Could you post a graph about the results of Reduction Matrix. A sentance about what was the idea would also be great.

**Comment by Raghwendra-Dey** [comment]
![Screenshot from 2020-03-20 00-22-17](https://user-images.githubusercontent.com/45457947/77652897-fb4a1280-6f94-11ea-92f0-1c2678d44af4.png)
What i did is, i made a matrix of 10x324 and filled it with random values form the normal distribution and multiplied that with the state list and normalized the result and returned the same.

**Comment by Gauraviitkgp** [comment]
> ![Screenshot from 2020-03-20 00-22-17](https://user-images.githubusercontent.com/45457947/77652897-fb4a1280-6f94-11ea-92f0-1c2678d44af4.png)
> What i did is, i made a matrix of 10x324 and filled it with random values form the normal distribution and multiplied that with the state list and normalized the result and returned the same.

Seems that the curve is rising because it is able to solve 1-2 step games and not 4 step ones.  So we need to tweak stuff. I get what you did, we could try training it obviously. Did you actually make a matrix o you inserted a layer of tensor-flow that acts as a matrix?

**Comment by Raghwendra-Dey** [comment]
> > ![Screenshot from 2020-03-20 00-22-17](https://user-images.githubusercontent.com/45457947/77652897-fb4a1280-6f94-11ea-92f0-1c2678d44af4.png)
> > What i did is, i made a matrix of 10x324 and filled it with random values form the normal distribution and multiplied that with the state list and normalized the result and returned the same.
> 
> Seems that the curve is rising because it is able to solve 1-2 step games and not 4 step ones. So we need to tweak stuff. I get what you did, we could try training it obviously. Did you actually make a matrix o you inserted a layer of tensor-flow that acts as a matrix?

I made a normal matrix (a numpy array) and just multiplied it, and normalized it using sklearn

**Comment by Gauraviitkgp** [comment]
That's not i think how you should approach it. So what you are currently doing is, 
S[1x324] * A[324,10] = S'[1x10] right?
Instead have a fully connected layer of size 10, with no bias added and non trainable weights. It does the same thing. And try applying softmax outputs for probability. Also, I think training the values would be much helpful. And remember that S' should be a one-hot integer rather than a probability distribution S' is not random vector its a deterministic vector

**Comment by Raghwendra-Dey** [comment]
> That's not i think how you should approach it. So what you are currently doing is,
> S[1x324] * A[324,10] = S'[1x10] right?
> Instead have a fully connected layer of size 10, with no bias added and non trainable weights. It does the same thing. And try applying softmax outputs for probability. Also, I think training the values would be much helpful. And remember that S' should be a one-hot integer rather than a probability distribution S' is not random vector its a deterministic vector

okay, got it.
But S was not a one hot vector right since since we are using 6(faces)x9(cubes in one face)x6(colors) dims vector right so this means S is not a one hot vector, so i think it would be wrong to force S' to be a one hot vector.

**Comment by Gauraviitkgp** [comment]
> okay, got it.
> But S was not a one hot vector right since since we are using 6(faces)x9(cubes in one face)x6(colors) dims vector right so this means S is not a one hot vector, so i think it would be wrong to force S' to be a one hot vector.

1) Yeah you are right i guess. I think better way would be to use Sigmoid as a the activation function, the sigmoid will tell you the probability of that bit being one. That way we have 2^10 independent states. So 

> _S'[x]=P(x=1|S)_ 

2. And first test it on a simple gridworld example before implementing it for Rubik. Push the gridworld test environment before testing on it. You can use this [environmen](https://github.com/Gauraviitkgp/Tensorflow/blob/master/2_RL/2-TestingPolicy.ipynb)t if you like
